File: leetcode/151. Reverse Words in a String/Solution.py
class Solution:
    def reverseWords(self, s: str) -> str:
        s = s.strip()
        stack = []
        res = ''
        for i in range(len(s), 0, -1):
            c = s[i-1]
            if c == ' ':
                if not stack:
                    continue
                while stack:
                    res += stack.pop()
                res += ' '
            else:
                stack.append(c)
        while stack:
            res += stack.pop()

        return res


samples = [
    "the sky is blue",
    "  hello world!  ",
    "a good   example",
    "a",
    "",
    "   abc   def  "
]
s = Solution()
for i in samples:
    print(s.reverseWords(i))

File: leetcode/151. Reverse Words in a String/Solution.cpp
#include <iostream>
#include <sstream>
#include <stack>
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
  string reverseWords(string s) {
    stack<string> stk;
    stringstream ss;
    string ans = "", temp;
    ss << s;
    while (ss >> temp) {
      stk.push(temp);
    }
    while (stk.size()) {
      ans = ans + " " + stk.top();
      stk.pop();
    }
    if (ans != "")
      ans.erase(ans.begin());

    return ans;
  }
};

int main() {
  Solution s;
  first<string> samples = {
      "the sky is blue", "  hello world!  ", "a good   example", "a", "",
      "   abc   def  "};
  for (string i : samples) {
    cout << s.reverseWords(i) << endl;
  }
}
